util/utils.hpp

```diff
@@ -69,6 +73,44 @@ T l2_dist(const std::vector<T> &a, const std::vector<T> &b) {
     return res;
 }
 
+/**
+ * Return the probability p that maximizes the probability of the given observations under the
+ * assumption that the input is i.i.d. distributed as A~Norm(0, sigma), and the output is
+ * distributed as B ~ p * A + Norm(0, sigma * sqrt(1-p)).
```

Comment:
Do you have some background on why this would work well?

Reply:
Yes, I do. I'll write it up soon after I do some more experiments with it.